### PR TITLE
Fixing CI + Python Bindings (Pkgs 16/N)

### PR DESCRIPTION
**Pull request by padentomasello**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.** Changes *must* be discussed.

**Original Issue**: [corresponding issue on Github]

*Note:* You can add `closes #[issue number]` to automatically close the issue that this PR resolves when it is merged.

### Summary
[Explain the details for this change and the problem that the pull request solves]

### Test Plan (required)
[steps by which you tested that your fix resolves the issue. These might include specific commands and configurations]

### CODE REVIEW

File: cmake/TestUtils.cmake [hunk 1]
```diff
@@ -48,6 +48,7 @@ function(build_test)
     PUBLIC
     ${GTEST_IMPORTED_TARGETS}
     ${build_test_LIBS}
+    pthread
```
Comment (jacobkahn) [inline cmake/TestUtils.cmake]: Not portable/doesn't work for libomp with MKL - see what we do [here](https://github.com/flashlight/flashlight/blob/c82d4d05e43c21702d57af70c8071282080da1e2/flashlight/fl/common/CMakeLists.txt#L34)

We should also isolate which symbol needs this and see why pthread isn't being linked - which test fails to build?

File: flashlight/pkg/vision/CMakeLists.txt [hunk 2]
```diff
@@ -2,7 +2,13 @@ cmake_minimum_required(VERSION 3.10)
 
 add_library(fl_pkg_vision "")
 
-target_link_libraries(fl_pkg_vision PUBLIC flashlight)
+target_link_libraries(fl_pkg_vision PUBLIC flashlight fl_lib_set)
+
+target_include_directories(
```
Comment (jacobkahn) [inline flashlight/pkg/vision/CMakeLists.txt]: nit: is this necessary for any of the pkgs? I would've thought this already works as per linking with the `flashlight` target. If it's necessary and I'm wrong, it can be left alone

File: flashlight/pkg/text/CMakeLists.txt [hunk 3]
```diff
@@ -3,7 +3,14 @@ cmake_minimum_required(VERSION 3.10)
 add_library(
   fl_pkg_text
   ""
-  ) 
+  )
+
+target_include_directories(
```
Comment (jacobkahn) [inline flashlight/pkg/text/CMakeLists.txt]: Ditto with the below with this being necessary (?)

File: flashlight/lib/text/test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (jacobkahn) [inline flashlight/lib/text/test/CMakeLists.txt]: nit: remove line at beginning of file

File: flashlight/lib/text/CMakeLists.txt [hunk 5]
```diff
@@ -21,3 +21,9 @@ include(${CMAKE_CURRENT_LIST_DIR}/dictionary/CMakeLists.txt)
 
 # tokenizer
 include(${CMAKE_CURRENT_LIST_DIR}/tokenizer/CMakeLists.txt)
+
+if(FL_BUILD_TESTS)
+  add_subdirectory(${CMAKE_CURRENT_LIST_DIR}/test)
+endif()
+
+
```
Comment (jacobkahn) [inline flashlight/lib/text/CMakeLists.txt]: nit: extra line

File: flashlight/fl/test/CMakeLists.txt [hunk 6]
```diff
@@ -8,7 +8,8 @@ build_test(SRC ${DIR}/common/DynamicBenchmarkTest.cpp LIBS ${LIBS})
 build_test(SRC ${DIR}/common/HistogramTest.cpp LIBS ${LIBS})
 build_test(SRC ${DIR}/common/LoggingTest.cpp LIBS ${LIBS})
 build_test(SRC ${DIR}/common/SerializationTest.cpp LIBS ${LIBS})
-build_test(SRC ${DIR}/common/SequentialBuilderTest.cpp LIBS ${LIBS})
+build_test(SRC ${DIR}/common/SequentialBuilderTest.cpp LIBS ${LIBS}
```
Comment (jacobkahn) [inline flashlight/fl/test/CMakeLists.txt]: Ditto with my previous comment on the PR where you moved this. I'm slightly impartial to this belonging in `pkg/common`

File: .circleci/config.yml [hunk 7]
```diff
@@ -334,9 +370,16 @@ jobs:
           fl_build_app_objdet: "OFF"
           fl_build_app_lm: "OFF"
       - test_flashlight_inside_nvidia_docker:
-          fl_test_apex_dir: "flashlight/lib/test"
+          test_name: "fl_pkg_speech"
```
Comment (jacobkahn) [inline .circleci/config.yml:434]: What's the CI wait time like on this particular job (`ubuntu2004_cuda11_1_docker_flashlight_pkgs`)? If we can parallelize across pkgs, that's well worth it and should cut down on the test time significantly

It'll also make it much easier to reason about which things have broken if one can identify which pkgs they may have broken

File: .circleci/config.yml [hunk 8]
```diff
@@ -95,11 +95,14 @@ commands:
             make -j$(nproc) && make install"
   test_flashlight_inside_nvidia_docker:
     parameters:
+      test_name:
+        type: string
+        default: ""
```
Comment (jacobkahn) [inline .circleci/config.yml:147]: nit: if you leave out the default, it'll be required, which might be what we want

File: CMakeLists.txt [hunk 9]
```diff
@@ -235,13 +240,59 @@ if (FL_BUILD_CORE)
...
+else()
+  cmake_dependent_option(FL_BUILD_APP_ASR "Build asr task for Flashlight" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+  cmake_dependent_option(FL_BUILD_APP_IMGCLASS "Build image classification app" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+  cmake_dependent_option(FL_BUILD_APP_OBJDET "Build object detection app" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+  cmake_dependent_option(FL_BUILD_APP_LM "Build lm task for Flashlight" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+endif ()
+
+cmake_dependent_option(FL_BUILD_ALL_PKGS "Set all Flashlight apps ON by default" OFF "FL_BUILD_CORE;FL_BUILD_CONTRIB;" OFF)
+
+# Set build options for pkgs depending on which apps you are build
```
Comment (jacobkahn) [inline CMakeLists.txt]: It might make sense to have all of the `cmake_dependent_option` calls for `pkg`s should also have the lib flags in their required options (see the comment on top of this review).

**Review by jacobkahn — COMMENTED** [review]
I see we have no options to gate building the lib components - this sort of defeats the purpose since it's all or nothing. Did you intend to add these? We should add them then force them on/off/throw errors (your choice as to behavior) in addition to modifying the `cmake_dependent_option` components.

File: flashlight/pkg/CMakeLists.txt [hunk 10]
```diff
@@ -1,18 +1,36 @@
 cmake_minimum_required(VERSION 3.10)
 
-set(FL_PKG_VISION_DIR ${FL_PKG_DIR}/vision)
-include(${FL_PKG_VISION_DIR}/CMakeLists.txt)
+if(FL_BUILD_PKG_RUNTIME)
+  set(FL_PKG_RUNTIME_DIR ${FL_PKG_DIR}/runtime)
+  include(${FL_PKG_RUNTIME_DIR}/CMakeLists.txt)
+  set(INSTALLABLE_TARGETS ${INSTALLABLE_TARGETS} fl_pkg_runtime)
+  setup_install_headers(${FL_PKG_RUNTIME_DIR} ${FL_INSTALL_INC_DIR_HEADER_LOC}/pkg)
```
Comment (jacobkahn) [inline flashlight/pkg/CMakeLists.txt:22]: Have you checked that the layout of the installed headers is correct given these changes? Probably fine, but worth checking

File: flashlight/app/CMakeLists.txt [hunk 11]
```diff
@@ -52,15 +32,12 @@ endif ()
 
 if (FL_BUILD_APP_OBJDET)
   message(STATUS "Building flashlight Object Detection app")
-  # TODO (padentomasello) revist if we can pull out some functionality from IMGCLASS
```
Comment (jacobkahn) [inline flashlight/app/CMakeLists.txt:86]: 👍

**Review by jacobkahn — CHANGES_REQUESTED** [review]
Submitted early. A few last comments:

File: cmake/TestUtils.cmake (hunk 1 above)
Comment (padentomasello) [reply cmake/TestUtils.cmake]: @jacobkahn All tests fails for any library which does not explicitly link to some Threading library, because gtest requires it. (It was getting linked transitively through the audio library before, but now that they are separate, I think we need to add explicitly, https://github.com/flashlight/flashlight/blob/master/flashlight/lib/audio/feature/CMakeLists.txt#L26-L32). Will follow the convention mentioned in the link.

File: flashlight/pkg/vision/CMakeLists.txt (hunk 2 above)
Comment (padentomasello) [reply flashlight/pkg/vision/CMakeLists.txt]: I was surprised by this as well. Let me double check and get back to you.

File: cmake/TestUtils.cmake (hunk 1 above)
Comment (jacobkahn) [reply cmake/TestUtils.cmake]: Sounds good. I've seen this before. That's fine.

File: flashlight/fl/test/CMakeLists.txt (hunk 6 above)
Comment (padentomasello) [reply flashlight/fl/test/CMakeLists.txt]: now in pkg/runtime

File: flashlight/pkg/CMakeLists.txt (hunk 10 above)
Comment (padentomasello) [reply flashlight/pkg/CMakeLists.txt:22]: Checked! Looks good to me.

File: .circleci/config.yml (hunk 8 above)
Comment (padentomasello) [reply .circleci/config.yml:147]: My intention was to not have it be required actually. Let me know if you think it should be, but it's only purpose is to add some information to the circle CI UI, so not really required functionally.

File: .circleci/config.yml (hunk 7 above)
Comment (padentomasello) [reply .circleci/config.yml:434]: The CI time for `ubuntu2004_cuda11_1_docker_flashlight_pkgs` is _just_ over `ubuntu2004_cuda11_1_docker_flashlight_fl_core` ( 26m vs 25 m ).  I agree with the benefits, and actually my first pass on this had each pkg as a separate CI job, but I thought it was a bit cumbersome having so many. Given the very slight time increase, I think I'd still prefer it as is, but let me know if you feel strongly.

**Comment by padentomasello** [comment]
@jacobkahn I went back and forth on what would be best. My initial thought was to automatically set the build flags based on the dependencies. (if FL_PKG_SPEECH is on set FL_LIB_AUDIO to be on since it's required). This actually goes against the current design, which is to use cmake_dependent_option and express the dependency in the other way. (if FL_LIB_AUDIO is not ON, the force FL_PKG_SPEECH to be OFF, even if the user set it to be on). I actually didn't like either of these options because I felt it was confusing to the user. (I set FL_PKG_SPEECH to be on, but it get's set to OFF somewhere in the build system without me knowing. Or I set FL_LIB_SPEECH to OFF and it get's set to ON. either way was somewhat surprising). 

Anyway, long story short, I ended up wrapping cmake_dependent_option with another function (fl_dependent_option), which checks for the proper dependencies and fails if you try to do something that breaks the dependency chain. It's slightly less autonomated, but is very explicit (no surprises). Let me know what you thing.

File: .circleci/config.yml (hunk 7 above)
Comment (jacobkahn) [reply .circleci/config.yml:434]: If there's no time increase, that's just fine, and we can leave it as is.

File: CMakeLists.txt [hunk 12]
```diff
@@ -52,11 +52,120 @@ option(FL_BUILD_EXPERIMENTAL "Build internal experimental components" OFF)
...
+
+# --------------------------- Core Options  ---------------------------
 option(FL_BUILD_CORE "Build flashlight core" ON)
-cmake_dependent_option(FL_BUILD_CONTRIB
-  "Build and link additional flashlight contrib assets." ON
-  "FL_BUILD_CORE" OFF)
+
+# Small utility function which wraps cmake_dependent options and throws an error if the user
+# is trying to build something without the proper dependencies.
+macro(FL_DEPENDENT_OPTION OPTION TEXT VAL DEPS FRCE)
```
Comment (jacobkahn) [inline CMakeLists.txt]: Would be great if you could do a few quick things to improve this:
1. Move this to `InternalUtils.cmake`. It should be available to all CMake files.
2. Use `function` instead of macro so we can use `cmake_parse_arguments` and named arguments ([example](https://github.com/flashlight/flashlight/blob/c82d4d05e43c21702d57af70c8071282080da1e2/cmake/HalideUtils.cmake#L10-L13), [docs](https://cmake.org/cmake/help/v3.10/command/cmake_parse_arguments.html)). CMake can implicitly cast anything to anything which results in tons of unintended bugs, and having named arguments is a lot clearer. Concretely, after this change, we want the call to look like:
```
fl_dependent_option(
  OPTION FL_BUILD_PKG_SPEECH
  TEXT "Build speech package for Flashlight"
  VAL "${FL_BUILD_ALL_PKGS}"
  DEPS "FL_BUILD_CORE;FL_BUILD_CONTRIB;FL_BUILD_LIB_AUDIO"
  FRCE OFF)
```
Which is much clearer/less error prone.

Optional, but we could also make `DEPS` take multiple things, as in:
```
  DEPS FL_BUILD_CORE FL_BUILD_CONTRIB FL_BUILD_LIB_AUDIO
```
then create the semicolon-separated string as needed for `cmake_dependent_option` so the user doesn't have to deal with the weird formatting. This would basically be
```
set(DEP_STRING "")
for (DEP in ITEMS ${DEPS})
  set(DEP_STRING "${DEP_STRING};${DEP}")
endforeach()
```

File: CMakeLists.txt [hunk 13]
```diff
@@ -52,11 +52,120 @@ option(FL_BUILD_EXPERIMENTAL "Build internal experimental components" OFF)
...
+  foreach(DEP IN ITEMS ${DEPS})
+    if(NOT ${${DEP}} AND ${OPTION} AND "${${OPTION}}")
+      message(FATAL_ERROR "${DEP} Required to build ${OPTION}")
+    endif()
+  endforeach()
+  cmake_dependent_option(${OPTION} ${TEXT} "${VAL}" "${DEPS}" ${FRCE})
+endmacro()
+
+fl_dependent_option(
```
Comment (jacobkahn) [inline CMakeLists.txt]: I'd like it if we moved these `fl_dependent_option` calls closer to the targets that they pertain to, sort of like the `cmake_dependent_option` calls we had before. If you move the `fl_dependent_option` macro into `InternalUtils.cmake`, it should be available everywhere.

File: cmake/TestUtils.cmake [hunk 14]
```diff
@@ -48,6 +49,7 @@ function(build_test)
     PUBLIC
     ${GTEST_IMPORTED_TARGETS}
     ${build_test_LIBS}
+     ${CMAKE_THREAD_LIBS_INIT}
```
Comment (jacobkahn) [inline cmake/TestUtils.cmake:52]: 👍

File: .circleci/config.yml [hunk 15]
```diff
@@ -90,16 +93,20 @@ commands:
                 -DFL_BUILD_APP_IMGCLASS=<< parameters.fl_build_app_imgclass >> \
                 -DFL_BUILD_APP_OBJDET=<< parameters.fl_build_app_objdet >> \
                 -DFL_BUILD_APP_LM=<< parameters.fl_build_app_lm >> \
-                -DBUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
+                -DFL_BUILD_ALL_PKGS=<< parameters.fl_build_all_pkgs >> \
+                -DFL_BUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
```
Comment (jacobkahn) [inline .circleci/config.yml]: `BUILD_SHARED_LIBS` is a standard CMake option -- I don't think we want this change
```suggestion
                -DBUILD_SHARED_LIBS=<< parameters.build_shared_libs >> \
```

**Review by jacobkahn — CHANGES_REQUESTED** [review]
`fl_dependent_option` is a nice abstraction and is a great way of expressing this. I like this a lot more. Just a few final comments with respect to cleanup.

One other thing that I'd appreciate your doing -- updating the documentation/build options and writing up how this works in the [README](https://github.com/flashlight/flashlight/blob/master/README.md) and [Dockerfile things](https://github.com/flashlight/flashlight/tree/master/.docker) if needed.

We also need to properly-gate dependencies in the [`flashlightConfig.cmake` template](https://github.com/flashlight/flashlight/blob/2081246aded0a0e2a7cb9c7fc965424b2af1f7be/cmake/flashlightConfig.cmake.in) based on these new options

File: CMakeLists.txt (hunk 12 above)
Comment (padentomasello) [reply CMakeLists.txt]: 1. Will definitely do. Good catch. 
2. The problem with using function is that the scope is not global, so we can't actually perform the check on the variables that are passed in because they only exist on the global scope. (And I think setting the option will also only be for the local scope, but not sure).

File: .circleci/config.yml (hunk 15 above)
Comment (padentomasello) [reply .circleci/config.yml]: Ah yes. Good catch. Unintentional.

File: CMakeLists.txt [hunk 16]
```diff
@@ -114,34 +109,42 @@ if (FL_BUILD_TESTS)
...
+  include(${FL_BINDING_PYTHON}/CMakeLists.txt)
+endif ()
+
+setup_install_headers(${FL_LIB_DIR} ${FL_INSTALL_INC_DIR_HEADER_LOC})
+
+# --------------------------- Core Options  ---------------------------
+
+fl_dependent_option(FL_BUILD_CORE "Build flashlight core" ON "FL_BUILD_LIB_COMMON" OFF)
+
+# Small utility function which wraps cmake_dependent options and throws an error if the user
```
Comment (jacobkahn) [inline CMakeLists.txt]: Nit: remove/move this comment to `InternalUtils.cmake`

File: cmake/flashlightConfig.cmake.in [hunk 17]
```diff
@@ -96,21 +93,46 @@ endif()
...
+endif()
+if (@FL_BUILD_LIB_AUDIO@)
+  set(flashlight_LIBRARIES ${flashlight_LIBRARIES} flashlight::fl_lib_audio)
+endif()
+if (@FL_BUILD_LIB_TEXT@)
+  set(flashlight_LIBRARIES ${flashlight_LIBRARIES} flashlight::fl_lib_text)
+endif()
+
+# core
+
```
Comment (jacobkahn) [inline cmake/flashlightConfig.cmake.in]: nit: remove newline

File: flashlight/lib/CMakeLists.txt [hunk 18]
```diff
@@ -1,17 +1,40 @@
 cmake_minimum_required(VERSION 3.10)
 
+# --------------------------- Lib Options  ---------------------------
+# Build all libs by default. In order to selectively build libs,
```
Comment (jacobkahn) [inline flashlight/lib/CMakeLists.txt]: I think this comment needs to be updated?

**Comment by padentomasello** [comment]
@jacobkahn Everything is opt in now. Documentation is fixed, and fixed dependencies in flashlightConfig.cmake. If these all look good to go, I'll merge it into one branch and import it to phabricactor to start working on some buck targets

File: README.md [hunk 19]
```diff
@@ -41,7 +41,7 @@ Flashlight is broken down into a few parts:
 - [**`flashlight/lib`**](flashlight/lib) contains kernels and standalone utilities for sequence losses, beam search decoding, text processing, and more.
 - [**`flashlight/fl`**](flashlight/fl) is the core neural network library using the [ArrayFire](https://github.com/arrayfire/arrayfire) tensor library.
 - [**`flashlight/app`**](flashlight/app) are applications of the core library to machine learning across domains.
-- [**`flashlight/ext`**](flashlight/ext) are extensions on top of Flashlight and ArrayFire that are useful across apps.
+- [**`flashlight/pkg**](flashlight/pkg) are packages built on top of Flashlight and ArrayFires for specific purposes (Domain libraries for speech and vision, halide extensions).
```
Comment (jacobkahn) [inline README.md]:
```suggestion
- [**`flashlight/pkg**](flashlight/pkg) are packages built on top of Flashlight and ArrayFire for specific purposes (Domain libraries for speech, vision, and text, and Halide extensions).
```

File: README.md [hunk 20]
```diff
@@ -420,12 +486,17 @@ The Flashlight CMake build accepts the following build options (prefixed with `-
 
 ### Building Your Own Project with Flashlight
 Flashlight is most-easily linked to using CMake. Flashlight exports the following CMake targets when installed:
-- `flashlight::fl-libraries` — contains flashlight libraries headers and symbols.
+- `flashlight::fl_lib_common` — contains flashlight libraries for common headers and symbols used throughout flashlight.
+- `flashlight::fl_lib_set` — contains flashlight libraries for headers and symbols pertaining to sets.
+- `flashlight::fl_lib_sequence` — contains flashlight libraries for headers and symbols pertaining to sequences.
+- `flashlight::fl_lib_audio` — contains flashlight libraries for headers and symbols pertaining to audio.
+- `flashlight::fl_lib_text` — contains flashlight libraries for headers and symbols pertaining to text.
 - `flashlight::flashlight` — contains flashlight libraries as well as the flashlight core autograd and neural network library.
```
Comment (jacobkahn) [inline README.md:494]: This is no longer correct (re `flashlight::flashlight`) - can we update it?

**Review by jacobkahn — COMMENTED** [review]
Awesome. Looks great. A few final quick things.

File: flashlight/lib/CMakeLists.txt (hunk 18 above)
Comment (padentomasello) [reply flashlight/lib/CMakeLists.txt]: Yep, good catch.

File: README.md (hunk 20 above)
Comment (padentomasello) [reply README.md:494]: We still have the `flashlight` library which is core. We no longer have fl-libraries.

**Review by jacobkahn — APPROVED** [review]
Looks good to me
